mne/preprocessing/artifact_detection.py

```diff
@@ -4,12 +4,105 @@
 
 
 import numpy as np
-
 from ..annotations import (Annotations, _annotations_starts_stops)
 from ..transforms import (quat_to_rot, _average_quats, _angle_between_quats,
                           apply_trans, _quat_to_affine)
+from ..filter import filter_data
 from .. import Transform
-from ..utils import (_mask_to_onsets_offsets, logger)
+from ..utils import (_mask_to_onsets_offsets, logger, verbose)
+
+
+@verbose
+def annotate_muscle_zscore(raw, threshold=4, picks=None, min_length_good=0.1,
+                           filter_freq=(110, 140), n_jobs=1, verbose=None):
+    """Detect segments that likely contain muscle artifacts.
+
+    Detects data segments containing activity in the frequency range given by
+    ``filter_freq``, that exceeds the specified z-score threshold.
+
+    Raw data is band-pass filtered between ``filter_freq`` especified
+    frequencies (default is 110 - 140 Hz), the signal envelope computed,
+    z-scored across samples, channel summation and division by the square root
+    of the channel number, and low-pass filtered to better capture beginning
+    and end of muscle activity and false positive transient peaks.
+
+    .. note::
+        Use a single channel type.
+
+    Parameters
+    ----------
+    raw : instance of Raw
+        Data to estimate segments with muscle artifacts.
+    threshold : float
+        The threshold in z-scores for marking segments as containing muscle
+        activity artifacts.
+    %(picks_all)s
+    min_length_good : float | None
+        The shortest allowed duration of "good data" (in seconds) between
+        adjacent annotations; shorter segments will be incorporated into the
+        surrounding annotations.``None`` is equivalent to ``0``.
+        Default is ``0.1``.
+    filter_freq : array-like, shape (2,)
+        The lower and upper frequencies of the band-pass filter.
+        Default is ``(110, 140)``.
+    %(n_jobs)s
+    %(verbose)s
+
+    Returns
+    -------
+    annot : mne.Annotations
+        Periods with muscle artifacts annotated as BAD_muscle.
+    scores_muscle : array
+        Z-score values averaged across channels for each sample.
+    """
+    from scipy.stats import zscore
+    from scipy.ndimage.measurements import label
+
+    raw_copy = raw.copy()
+    raw_copy.pick(picks)
+    ch_type = raw_copy.get_channel_types()
+
+    # Remove ref chans if MEG data just in case
+    meg = [True for e in ['mag', 'grad'] if (e in ch_type)]
+    if bool(meg) is True:
+        raw_copy.pick_types(ref_meg=False)
+        ch_type = raw_copy.get_channel_types()
+
+    # Only one type of channel, otherwise z-score will be biased
+    assert(len(np.unique(ch_type)) == 1), \
+        'Different channel types, pick one type'
```

Comment:
(nitpick) sure, `np.unique` will work, just like `set` worked before.  But I think you missed the point of my earlier comment. I was saying that if you rebased this branch, so that the new changes to `get_channel_types` were available, then you could avoid using *anything* here (neither `set` nor `np.unique`), because you could simply pass `unique=True` to the earlier calls to `get_channel_types`. This should be fine because both places where you use `ch_type` are places where it doesn't matter which *specific channel* has which *specific type*... rather, all you care about is whether more than 1 type is present, and whether that type is a MEG type or not.

Reply:
Quite separately from the nitpick, and something I should have mentioned before, is that we don't typically use `assert` statements for things that are actually likely to happen when users use the function. Instead, we tend to raise more informative error types (here, probably `ValueError` is closest?) and give more narrative error messages like "annotate_muscle_zscore only works on one channel type at a time, use one of the picking methods prior to calling this function" (or something like that).  And anytime we raise an error, we add a test to make sure the error *actually* gets raised  when we give bad inputs to the function. `assert` statements we generally use in our tests, or in rare cases in the codebase where we don't expect something to ever occur, but if it did, then it would be really bad / hard to detect based on later output.

Reply:
+1, `ValueError` would be fine (I think I also suggested it in this PR, but the comments no longer show).

Reply:
OK. I changed that, now ``ch_type`` is a parameter and this section has been removed